### PR TITLE
Filter arguments for `chat api-listen`

### PR DESCRIPTION
**Pull request by zapu**
Experimental filtering for `api-listen`, for chat messages. It's at the command level, not at the service level, so the messages still go through RPC layer, but they are not printed to consumer.

Example usage:
```
keybase chat api-listen --filter-channels '[{"name":"tuserac88e1@rooter,tuser5ca897,tuserac88e0" }, {"name":"ttca897", "topic_name": "core", "members_type": "team"}]'
keybase chat api-listen --filter-channel '{"name":"tuserac88e1@rooter,tuser5ca897,tuserac88e0" }'
```

### CODE REVIEW

File: go/client/chat_api_listen_display.go [hunk 1]
```diff
@@ -98,6 +121,42 @@ func (d *chatNotificationDisplay) formatMessage(inMsg chat1.IncomingMessage) *Me
...
+func (d *chatNotificationDisplay) matchFilters(conv *chat1.InboxUIItem) bool {
+	if len(d.filtersNormalized) == 0 {
+		// No fliters - every message is relayed.
+		return true
+	}
+	for _, v := range d.filtersNormalized {
+		if matchMembersTypeToFilter(v, conv.MembersType) && v.Name == strings.ToLower(conv.Name) {
+			// If any of the following were specified by user and differ from
+			// what was received, filter the msg out.
+			if v.TopicType != "" && conv.TopicType.String() != v.TopicType {
```
Comment (joshblum) [inline go/client/chat_api_listen_display.go]: why not combine these checks into the conditional as well?

File: go/client/chat_api_listen_display.go [hunk 2]
```diff
@@ -98,6 +121,42 @@ func (d *chatNotificationDisplay) formatMessage(inMsg chat1.IncomingMessage) *Me
...
+		default:
+			return false
+		}
+	}
+	return filter.MembersType == typ.String()
+}
+
+func (d *chatNotificationDisplay) matchFilters(conv *chat1.InboxUIItem) bool {
+	if len(d.filtersNormalized) == 0 {
+		// No fliters - every message is relayed.
```
Comment (joshblum) [inline go/client/chat_api_listen_display.go]:
```suggestion
		// No filters - every message is relayed.
```

File: go/client/cmd_chat_api_listen.go [hunk 3]
```diff
@@ -79,13 +118,29 @@ func sendPing(cli keybase1.SessionClient) error {
...
 	}
 
 	chatDisplay := newChatNotificationDisplay(c.G(), c.showLocal, c.hideExploding)
+
+	if err := chatDisplay.setupFilters(context.TODO(), c.channelFilters); err != nil {
+		return err
+	}
+
+	if filterLen := len(chatDisplay.filtersNormalized); filterLen > 0 {
+		c.ErrWriteLn("Message filtering is active: %d filters", filterLen)
```
Comment (joshblum) [inline go/client/cmd_chat_api_listen.go]:
```suggestion
		c.ErrWriteLn("Message filtering is active with %d filters", filterLen)
```

File: go/client/chat_api_listen_display.go (hunk 1 above)
Comment (zapu) [reply go/client/chat_api_listen_display.go]: so fold all checks into one ifs? or extract to separate function or to `matchMembersTypeToFilter` (probably renamed then)? thanks

Comment (joshblum) [reply go/client/chat_api_listen_display.go]: whichever you think makes it most readable, probably putting them all into one `if` and putting each condition on a new line is easiest

**Pull request merged**